cmd/release-controller/sync_release_payload.go

```diff
@@ -0,0 +1,144 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"github.com/openshift/release-controller/pkg/apis/release/v1alpha1"
+	releasecontroller "github.com/openshift/release-controller/pkg/release-controller"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/klog/v2"
+	"sort"
+)
+
+func (c *Controller) processReleasePayloadEvent(obj interface{}) {
+	switch t := obj.(type) {
+	case *v1alpha1.ReleasePayload:
+		klog.Infof("Processing ReleasePayload Event for: %s/%s", t.Namespace, t.Name)
+		c.addReleasePayloadQueueKey(queueKey{namespace: t.Namespace, name: t.Name})
+	default:
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %T", obj))
+	}
+}
+
+func (c *Controller) syncReleasePayload(key queueKey) error {
+	// Get the ReleasePayload resource with this namespace/name
+	payload, err := c.releasePayloadLister.ReleasePayloads(key.namespace).Get(key.name)
+	if err != nil {
+		// The ReleasePayload resource may no longer exist, in which case we stop processing.
+		if errors.IsNotFound(err) {
+			utilruntime.HandleError(fmt.Errorf("payload '%s/%s' in work queue no longer exists", key.namespace, key.name))
+			return nil
+		}
+		return err
+	}
+	klog.Infof("Syncing ReleasePayload: '%s/%s'", payload.Namespace, payload.Name)
+
+	// This is the entry point, for the release-controller, to respond to any updates to ReleasePayloads as they occur...
+
+	// TODO: Add business logic here
+
+	return nil
+}
+
+func (c *Controller) ensureReleasePayload(release *releasecontroller.Release, tagName string) (*v1alpha1.ReleasePayload, error) {
+	payload, err := c.releasePayloadLister.ReleasePayloads(release.Target.Namespace).Get(tagName)
+	if err == nil {
+		klog.V(4).Infof("ReleasePayload: %s/%s already exists", payload.Namespace, payload.Name)
+		return payload, nil
+	}
+	if !errors.IsNotFound(err) {
+		return nil, err
+	}
+	payload = newReleasePayload(release, tagName, c.jobNamespace, c.prowNamespace)
+	payload, err = c.releasePayloadClient.ReleasePayloads(release.Target.Namespace).Create(context.TODO(), payload, metav1.CreateOptions{})
+	if err == nil {
+		klog.V(4).Infof("ReleasePayload: %s/%s created", payload.Namespace, payload.Name)
+		return payload, nil
+	}
+	if !errors.IsAlreadyExists(err) {
+		return nil, err
+	}
+	// perform a live lookup if we are racing to create the payload
+	klog.V(4).Infof("Performing live lookup of ReleasePayload: %s/%s", payload.Namespace, payload.Name)
+	return c.releasePayloadClient.ReleasePayloads(release.Target.Namespace).Get(context.TODO(), tagName, metav1.GetOptions{})
+}
+
+func newReleasePayload(release *releasecontroller.Release, name, jobNamespace, prowNamespace string) *v1alpha1.ReleasePayload {
+	payload := v1alpha1.ReleasePayload{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: release.Target.Namespace,
+			Labels: map[string]string{
```

Comment:
remind me why we have these?  they look like information duplicated from actual field.  Sometimes we need this for selection capability, but I don't remember why these fields need to be labels.

Reply:
I was wondering the safe thing, and err'd on the side of "they must be there for a reason" and included them.  I believe they came from a discussion about selection criteria, possibly before we agreed that the name of the release and the releasepayload would be the same...maybe?  I can certainly remove them since they are not currently being used by any of the logic.

Reply:
yes, better to remove these and add them later if we need them.